Trakt/Api/TraktApi.cs

```diff
@@ -923,46 +835,131 @@ private async Task<TraktSyncResponse> SendEpisodePlaystateUpdatesInternalAsync(I
             }
         }
 
-        public async Task RefreshUserAuth(TraktUser traktUser)
+        public string AuthorizeDevice(TraktUser traktUser)
         {
-            var data = new TraktUserTokenRequest
+            var deviceCodeRequest = new
             {
-                client_id = TraktUris.Id,
-                client_secret = TraktUris.Secret,
-                redirect_uri = "urn:ietf:wg:oauth:2.0:oob"
+                client_id = TraktUris.ClientId
             };
-
-            if (!string.IsNullOrWhiteSpace(traktUser.PIN))
+            
+            TraktDeviceCode deviceCode;
+            using (var response = PostToTrakt(TraktUris.DeviceCode, deviceCodeRequest, null))
             {
-                data.code = traktUser.PIN;
-                data.grant_type = "authorization_code";
+                deviceCode = _jsonSerializer.DeserializeFromStream<TraktDeviceCode>(response.Result);
             }
-            else if (!string.IsNullOrWhiteSpace(traktUser.RefreshToken))
+
+            // Start polling in the background
+            Plugin.Instance.PollingTasks[traktUser.LinkedMbUserId] = Task.Run(() => PollForAccessToken(deviceCode, traktUser));
+            
+            return deviceCode.user_code;
+        }
+
+        public async Task<bool> PollForAccessToken(TraktDeviceCode deviceCode, TraktUser traktUser)
+        {
+            var deviceAccessTokenRequest = new
             {
-                data.refresh_token = traktUser.RefreshToken;
-                data.grant_type = "refresh_token";
+                code = deviceCode.device_code,
+                client_id = TraktUris.ClientId,
+                client_secret = TraktUris.ClientSecret
+            };
+
+            var pollingInterval = deviceCode.interval;
+            var expiresAt = DateTime.UtcNow.AddSeconds(deviceCode.expires_in);
+            _logger.LogInformation("Polling for access token every {PollingInterval}s. Expires at {ExpiresAt} UTC.", pollingInterval, expiresAt);
+            while (DateTime.UtcNow < expiresAt)
+            {
+                try
+                {
+                    using (var response = await PostToTrakt(TraktUris.DeviceToken, deviceAccessTokenRequest).ConfigureAwait(false))
+                    {
+                        _logger.LogInformation("Device successfully authorized");
+                        
+                        var deviceAccessToken = _jsonSerializer.DeserializeFromStream<TraktDeviceAccessToken>(response.Content);
+                        if (deviceAccessToken != null)
+                        {
+                            traktUser.AccessToken = deviceAccessToken.access_token;
+                            traktUser.RefreshToken = deviceAccessToken.refresh_token;
+                            traktUser.AccessTokenExpiration = DateTime.Now.AddMonths(2);
```

Comment:
can we extract this `2` magic months count to some named constant?

Reply:
I've changed it a bit